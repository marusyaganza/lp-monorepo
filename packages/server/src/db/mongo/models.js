const { formatIds, formatFilter } = require('../../utils/dbUtils');
const { ObjectId } = require('mongodb');

function createModel(db, table) {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result = await db.collection(table).find(formatFilter(filter));
    const data = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields) {
    const data = { ...fields };
    delete data.id;
    delete data.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    return result?.result;
  }

  async function deleteOne(filter) {
    console.log('filter', filter);
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    return result?.result;
  }

  async function createMany(entries) {
    const formattedEntries = entries.map(entry => ({
      ...entry,
      createdAt: Date.now()
    }));
    const result = await db.collection(table).insertMany(formattedEntries);
    return result.ops.map(formatIds);
  }

  return {
    findOne,
    findMany,
    createOne,
    createMany,
    updateOne,
    deleteOne
  };
}

module.exports = createModel;
