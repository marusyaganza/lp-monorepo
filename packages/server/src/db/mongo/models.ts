import { formatIds, formatFilter } from '../../utils/dbUtils';

import { Table, ModelType, DbData } from '..';

const { ObjectId } = require('mongodb');

function createModel<T extends DbData>(db: any, table: Table): ModelType<T> {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result: T[] = await db.collection(table).find(formatFilter(filter));
    const data: T[] = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields: Partial<T>) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields: Partial<T>) {
    const data = { ...fields };
    delete data?.id;
    delete data?.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    const isOk = result?.result?.ok ? true : false;
    const modifiedCount = result?.result?.nModified;
    return { ok: isOk, modifiedCount };
  }

  async function deleteOne(filter: Partial<T>) {
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    const isOk = result?.result?.ok ? true : false;
    return { ok: isOk };
  }

  return {
    findOne,
    findMany,
    createOne,
    updateOne,
    deleteOne
  };
}

export default createModel;
