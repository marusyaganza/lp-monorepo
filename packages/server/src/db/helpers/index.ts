export * from './sortingHelpers/sortingHelpers';
export * from './general/general';
