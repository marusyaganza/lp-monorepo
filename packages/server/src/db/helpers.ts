import { Document, Types } from 'mongoose';
import {
  Maybe,
  Language,
  Game,
  SortBy,
  SortWordsBy
} from '../generated/graphql';

type ModelDataType = {
  id?: Maybe<string>;
  user?: string;
};

interface DbData extends ModelDataType {
  _id?: string;
}

type DocumentType<T> =
  | (Document<unknown, unknown, T> &
      T & {
        _id: Types.ObjectId;
      })
  | null;

export interface WordsFilterType {
  sortBy?: SortBy | SortWordsBy | 'updatedAt';
  language: Language;
  isReverseOrder?: boolean;
  timesToLearn?: number | null;
  gameType?: Game;
  user?: string;
  tags?: string[];
  searchQuery?: string;
}

export interface WordsWithPaginationFilter extends WordsFilterType {
  page?: number;
  limit?: number;
}

export function formatData<T>(data: DocumentType<T>): T | null {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter?.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
const NO_TAGS_ID = '000000000000000000000000';

export function getWordsFilters(filter: WordsFilterType) {
  const {
    user,
    language = Language.English,
    sortBy,
    isReverseOrder,
    gameType,
    tags,
    searchQuery,
    timesToLearn = 5
  } = filter;

  // filter for a specific game
  let gameFilter;
  if (gameType === Game.Conjugation) {
    gameFilter = { conjugation: { $ne: null } };
  }
  if (gameType === Game.Audio) {
    gameFilter = { audioUrl: { $ne: '' } };
  }

  // filter based on tags
  let tagsFilters: { tags: string }[] = [];
  if (tags?.length) {
    tagsFilters = tags
      ?.filter(tag => tag !== NO_TAGS_ID)
      .map(tag => {
        return { tags: tag };
      });
  }

  let noTagsFilter;
  if (tags?.includes(NO_TAGS_ID)) {
    noTagsFilter = { $or: [{ tags: [] }, { tags: null }] };
  }

  let searchFilter;
  if (searchQuery) {
    const searchRegexp = { $regex: new RegExp(searchQuery, 'i') };
    searchFilter = {
      $or: [
        { name: searchRegexp },
        { 'defs.def': searchRegexp },
        { alternativeSpelling: searchRegexp },
        { stems: searchRegexp }
      ]
    };
  }

  // filter based on learning status
  let learningStatusFilters: Record<string, unknown>[] = [];
  if (gameType && sortBy !== SortBy.MemoryRefresher) {
    learningStatusFilters = [
      { isLearned: { $ne: true } },
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $lt: timesToLearn } },
          { [`statistics.${gameType}.successRate`]: null }
        ]
      }
    ];
  }
  if (sortBy === SortBy.MemoryRefresher) {
    learningStatusFilters = [
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $gte: timesToLearn } },
          { isLearned: true }
        ]
      }
    ];
  }

  //sorting
  let orderNum = isReverseOrder ? -1 : 1;

  // words with the highest number of errors should go first
  if (sortBy === SortBy.ErrorCount) {
    orderNum = -1 * orderNum;
  }

  let sort: Record<string, number> = { $natural: orderNum };

  if (sortBy) {
    let propName: string = sortBy;
    if (gameType) {
      propName = `statistics.${gameType}.${sortBy}`;
    }

    if (sortBy !== SortBy.MemoryRefresher) {
      sort = { [propName]: orderNum };
    }
  }

  const filtersArray = [
    gameFilter,
    noTagsFilter,
    searchFilter,
    ...tagsFilters,
    ...learningStatusFilters
  ].filter(Boolean);

  const filters: Record<string, unknown> = {
    user,
    language
  };

  if (filtersArray?.length) {
    filters.$and = filtersArray;
  }

  return {
    sort,
    filters
  };
}
