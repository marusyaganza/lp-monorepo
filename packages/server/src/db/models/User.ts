import { User, UserType } from '../schema/User';
import { SignUpInput, Role } from '../../generated/graphql';
import { formatData, formatFilter } from '../helpers';

export interface UserModelType {
  findOne: (filter: Partial<UserType>) => Promise<UserType | null>;
  findMany: (filter: Partial<UserType>) => Promise<UserType[] | null>;
  createOne: (fields: SignUpInput & { role: Role }) => Promise<UserType | null>;
  updateOne: (
    fields: Partial<UserType>
  ) => Promise<{ ok: boolean; value: UserType | null }>;
  deleteOne: (filter: { id: string }) => Promise<{ ok: boolean }>;
}

export const UserModel: UserModelType = {
  async findOne(filter = {}) {
    const user = await User.findOne(formatFilter(filter));
    return formatData(user);
  },

  //TODO check this method when admin auth is ready
  async findMany(filter = {}) {
    const users = await User.find(formatFilter(filter));
    return users;
  },

  async createOne(fields) {
    const createdAt = Date.now();
    const user = await User.create({ ...fields, createdAt });
    return formatData(user);
  },

  //TODO check this method when GQL for this is ready
  async updateOne(fields) {
    const update = { ...fields };
    delete update.id;
    const { ok, value } = await User.findByIdAndUpdate(fields.id, update, {
      includeResultMetadata: true,
      new: true
    });
    return { ok: Boolean(ok), value: formatData(value) };
  },

  //TODO check this method when admin auth is ready
  async deleteOne(filter) {
    const result = { ok: false };
    const { deletedCount, acknowledged } = await User.deleteOne(
      formatFilter(filter)
    );
    if (acknowledged && deletedCount == 1) {
      result.ok = true;
    }
    return result;
  }
};
