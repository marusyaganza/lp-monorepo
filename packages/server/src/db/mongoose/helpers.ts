import { Types, Document } from 'mongoose';
import { User, Word } from '../../generated/graphql';
import { DbData } from 'db';

type DocumentType =
  | (Document<unknown, Record<string, any>, User | Word> &
      Omit<
        | User
        | (Word & {
            _id: Types.ObjectId;
          }),
        never
      >)
  | null;

export function formatData(data: DocumentType) {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
