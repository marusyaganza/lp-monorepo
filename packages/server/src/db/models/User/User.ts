import { User, UserType } from '../../schema/User';
import { SignUpInput } from '../../../generated/graphql';
import {
  AuthenticationError,
  OperationResolutionError
} from '../../../utils/apolloCustomErrors';
import { DEFAULT_ROLE } from '../../../constants/defaultValues';
import { hashPassword } from '../../../auth';
import { ERROR_MESSAGES } from '../../../constants/errorMessages';

export interface UserModelType {
  findOne: (filter: Partial<UserType>) => Promise<UserType | null | undefined>;
  createOne: (fields: SignUpInput) => Promise<UserType | null>;
  createTempUser: () => Promise<UserType | null>;
}

export const UserModel: UserModelType = {
  async findOne(filter = {}) {
    const { id } = filter;
    let user;
    if (id) {
      user = await User.findById(id);
    } else {
      user = await User.findOne(filter);
    }
    return user?.toObject();
  },

  async createOne(fields) {
    const { email, password } = fields;
    const existing = await User.findOne({ email });
    if (existing) {
      throw new AuthenticationError(`user with email ${email} already exists`);
    }
    const role = DEFAULT_ROLE;
    const hashedPassword = await hashPassword(password);
    if (!hashedPassword) {
      throw new OperationResolutionError(ERROR_MESSAGES.SIGN_UP_FAILED);
    }
    const user = await User.create({
      ...fields,
      role,
      password: hashedPassword
    });
    return user.toObject();
  },
  async createTempUser() {
    const role = DEFAULT_ROLE;
    const timestamp = Date.now();
    const firstName = 'Sam';
    const lastName = 'Smith';
    const email = `${firstName.toLowerCase()}.${lastName.toLowerCase()}.${timestamp}@example.com`;
    const hashedPassword = await hashPassword(timestamp.toString());
    if (!hashedPassword) {
      throw new OperationResolutionError(ERROR_MESSAGES.SIGN_UP_FAILED);
    }
    const user = await User.create({
      firstName,
      lastName,
      email,
      role,
      password: hashedPassword
    });
    return user.toObject();
  }
};
