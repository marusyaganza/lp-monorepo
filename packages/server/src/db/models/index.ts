import { UserModel, UserModelType } from './User';
import { WordModel, WordModelType } from './Word';
import { WordTagModel, WordTagModelType } from './WordTag';

export interface ModelsType {
  User: UserModelType;
  Word: WordModelType;
  WordTag: WordTagModelType;
}

export const models: ModelsType = {
  User: UserModel,
  Word: WordModel,
  WordTag: WordTagModel
};
