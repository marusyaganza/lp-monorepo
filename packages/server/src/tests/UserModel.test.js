import { UserModel } from '../db/models/User';
import { connectToDb, disconnectFromDb, dropDb } from './helpers';
import { testData } from './mocks/dbTestData';

const snapshotConfig = {
  createdAt: expect.any(String),
  id: expect.any(String),
  _id: expect.any(Object)
};

describe('User', () => {
  beforeAll(async () => {
    await connectToDb();
    await dropDb();
  });

  afterEach(async () => {
    await dropDb();
  });

  afterAll(async () => {
    await disconnectFromDb();
  });

  test('createOne', async () => {
    const result = await UserModel.createOne(testData.createUserInput);
    expect(result).toMatchSnapshot(snapshotConfig);
  });

  test('findOne', async () => {
    const user = await UserModel.createOne(testData.createUserInput);
    const result = await UserModel.findOne({
      email: testData.createUserInput.email
    });
    const result2 = await UserModel.findOne({ id: user.id });
    expect(result2).not.toBeNull();
    expect(result).toMatchSnapshot(snapshotConfig);
  });
  test('findMany', async () => {
    await UserModel.createOne(testData.createUserInput);
    await UserModel.createOne(testData.createUserInput);
    const result = await UserModel.findMany({ role: 'MEMBER' });
    expect(result).toHaveLength(2);
  });

  test('updateOne', async () => {
    const user = await UserModel.createOne(testData.createUserInput);
    const result = await UserModel.updateOne({
      ...testData.updateUserInput,
      id: user.id
    });
    expect(result.ok).toBe(true);
    const searchTResult = await UserModel.findOne({ id: user.id });
    expect(result.value).toEqual(searchTResult);
    expect(searchTResult).toMatchSnapshot(snapshotConfig);
  });

  test('deleteOne', async () => {
    const user1 = await UserModel.createOne(testData.createUserInput);
    await UserModel.createOne(testData.createUserInput);
    const result = await UserModel.deleteOne({ id: user1.id });
    expect(result).toEqual({ ok: true });
    const searchTResult = await UserModel.findOne({ id: user1.id });
    expect(searchTResult).toBeNull();
  });

  test('deleteOne if user does not exist', async () => {
    const user1 = await UserModel.createOne(testData.createUserInput);
    await UserModel.createOne(testData.createUserInput);
    const result = await UserModel.deleteOne({ email: 'test77@test.com' });
    expect(result).toEqual({ ok: false });
    const searchTResult = await UserModel.findOne({ id: user1.id });
    expect(searchTResult).not.toBe(null);
  });
});
