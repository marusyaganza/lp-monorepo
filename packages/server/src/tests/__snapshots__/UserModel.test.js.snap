// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User createOne 1`] = `
{
  "_id": Any<Object>,
  "createdAt": Any<String>,
  "email": "test@test.com",
  "firstName": "Test",
  "id": Any<String>,
  "lastName": "User1",
  "password": "password1",
  "primaryLanguage": "English",
  "role": "MEMBER",
}
`;

exports[`User findOne 1`] = `
{
  "_id": Any<Object>,
  "createdAt": Any<String>,
  "email": "test@test.com",
  "firstName": "Test",
  "id": Any<String>,
  "lastName": "User1",
  "password": "password1",
  "primaryLanguage": "English",
  "role": "MEMBER",
}
`;

exports[`User updateOne 1`] = `
{
  "_id": Any<Object>,
  "createdAt": Any<String>,
  "email": "test@test.com",
  "firstName": "Test3",
  "id": Any<String>,
  "lastName": "User1",
  "password": "password1",
  "primaryLanguage": "French",
  "role": "MEMBER",
}
`;
